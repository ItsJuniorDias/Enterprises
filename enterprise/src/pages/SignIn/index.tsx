/* eslint-disable no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable react/prop-types */
import React, { useState, useCallback, useRef } from 'react';
import { Image, View, TouchableOpacity, TextInput, Alert } from 'react-native';
import { Form } from '@unform/mobile';
import { FormHandles } from '@unform/core';
import { useNavigation } from '@react-navigation/core';

import { useDispatch } from 'react-redux';
import * as Yup from 'yup';
import { loginToAuthRequest } from '../../store/modules/auth/actions';

import getValidationErrors from '../../utils/getValidationErrors';

import api from '../../services/api';

import Input from '../../components/Input';
import Button from '../../components/Button';
import Loading from '../../components/Loading';

import iconGoogle from '../../assets/Google.png';
import iconFacebook from '../../assets/Facebook.png';
import forgotPassword from '../../assets/round-arrow_right_alt-24px.png';

import {
  Container,
  Title,
  Header,
  Body,
  Footer,
  Forgot,
  TextForgot,
  TextFooter,
  Row,
} from './styles';

interface SignInFormData {
  email: string;
  password: string;
}

const SignIn: React.FC = () => {
  const [loading, setLoading] = useState(false);

  const dispatch = useDispatch();

  const navigation = useNavigation();

  const formRef = useRef<FormHandles>(null);
  const passwordInputRef = useRef<TextInput>(null);

  const handleSignIn = useCallback(async (data: SignInFormData) => {
    try {
      formRef.current?.setErrors({});

      const schema = Yup.object().shape({
        email: Yup.string()
          .required('E-mail obrigatório')
          .email('Digite um e-mail válido'),
        password: Yup.string().required('Senha obrigatória'),
      });

      await schema.validate(data, {
        abortEarly: false,
      });

      setLoading(true);

      await api.post('/users/auth/sign_in', data);
      dispatch(loginToAuthRequest(data));

      setLoading(false);

      navigation.navigate('Home');
    } catch (err) {
      if (err instanceof Yup.ValidationError) {
        const errors = getValidationErrors(err);

        formRef.current?.setErrors(errors);
        return;
      }

      setLoading(false);

      Alert.alert(
        'Erro na autenticação',
        'Ocorreu um erro ao fazer login, cheque as credenciais',
      );
    }
  }, []);

  return (
    <Container>
      {loading && <Loading />}

      {!loading && (
        <View>
          <Header>
            <Title>Login</Title>
          </Header>

          <Body>
            <Form ref={formRef} onSubmit={handleSignIn}>
              <Input
                autoCorrect={false}
                autoCapitalize="none"
                keyboardType="email-address"
                name="email"
                icon="mail"
                placeholder="E-mail"
                returnKeyType="next"
                onSubmitEditing={() => {
                  passwordInputRef.current?.focus();
                }}
              />

              <Input
                ref={passwordInputRef}
                secureTextEntry
                name="password"
                icon="lock"
                placeholder="Senha"
                returnKeyType="send"
                onSubmitEditing={() => {
                  formRef.current?.submitForm();
                }}
              />

              <Forgot>
                <TextForgot>Esqueceu sua senha ?</TextForgot>
                <Image source={forgotPassword} />
              </Forgot>

              <Button
                onPress={() => {
                  formRef.current?.submitForm();
                }}
              >
                Entrar
              </Button>
            </Form>
          </Body>

          <Footer>
            <View>
              <TextFooter>Ou faça login com sua conta</TextFooter>
            </View>

            <Row>
              <TouchableOpacity>
                <Image source={iconGoogle} />
              </TouchableOpacity>

              <TouchableOpacity>
                <Image source={iconFacebook} />
              </TouchableOpacity>
            </Row>
          </Footer>
        </View>
      )}
    </Container>
  );
};

export default SignIn;
