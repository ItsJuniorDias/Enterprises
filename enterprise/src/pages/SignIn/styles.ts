import styled from 'styled-components/native';

export const Container = styled.View`
  flex: 1;
  padding: 0 14px;
`;

export const Title = styled.Text`
  font-size: 28px;
  color: #0d253c;
  font-family: 'Poppins-SemiBold';
`;

export const Header = styled.View`
  margin-top: 32px;
`;

export const Body = styled.View`
  margin-top: 73px;
`;

export const Forgot = styled.TouchableOpacity`
  width: 100%;
  flex-direction: row;
  justify-content: flex-end;
  margin-top: 8px;
`;

export const TextForgot = styled.Text`
  font-size: 14px;
  color: #2d4379;
  font-family: 'Poppins-Regular';
  margin-right: 8px;
`;

export const TextFooter = styled.Text`
  font-size: 14px;
  color: #2d4379;
  font-family: 'Poppins-Regular';
  margin-top: -8px;
  margin-bottom: 8px;
`;

export const ContentFooter = styled.View`
  background-color: red;
  height: 80px;
  justify-content: flex-start;
`;

export const Row = styled.View`
  flex-direction: row;
`;

export const Footer = styled.View`
  margin-top: 135px;
  width: 100%;
  align-items: center;
  justify-content: center;
`;
