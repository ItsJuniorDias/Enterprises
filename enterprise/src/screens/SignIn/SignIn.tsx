import React, { useState, useCallback, useRef } from 'react';
import { TextInput, Alert, Image, TouchableOpacity } from 'react-native';
import { Form } from '@unform/mobile';
import { FormHandles } from '@unform/core';
import { useNavigation } from '@react-navigation/native';
import { useDispatch } from 'react-redux';
import * as Yup from 'yup';
import { loginToAuthRequest } from '../../store/modules/auth/actions';
import { getValidationErrors } from '../../utils/getValidationErrors';
import api from '../../services/api';
import Input from '../../components/Input/Input';
import { Button, Loading } from '../../components';

import { useAuth } from '../../hooks';

import iconGoogle from '../../assets/Google.png';
import iconFacebook from '../../assets/Facebook.png';
import forgotPassword from '../../assets/round-arrow_right_alt-24px.png';

import {
  Container,
  Title,
  Header,
  Body,
  TouchableText,
  Text,
  Footer,
  TextFooter,
  Row,
} from './styles';

interface SignInFormData {
  email: string;
  password: string;
}

export const SignIn = () => {
  const testIDs = useRef({
    input_email: 'inputEmail_testID',
    input_password: 'inputPassword_testID',
    button: 'buttonSignIn_testID',
  }).current;

  const [loading, setLoading] = useState(false);

  const dispatch = useDispatch();
  const navigation = useNavigation();

  const { login } = useAuth();

  const formRef = useRef<FormHandles>(null);
  const passwordInputRef = useRef<TextInput>(null);

  const handleSignIn = useCallback(async (data: SignInFormData) => {
    try {
      formRef.current?.setErrors({});

      const schema = Yup.object().shape({
        email: Yup.string()
          .required('E-mail obrigatório')
          .email('Digite um e-mail válido'),
        password: Yup.string().required('Senha obrigatória'),
      });

      await schema.validate(data, {
        abortEarly: false,
      });

      setLoading(true);

      await login(data);

      // dispatch(loginToAuthRequest(data));

      setLoading(false);

      navigation.navigate('/Home');
    } catch (err) {
      if (err instanceof Yup.ValidationError) {
        const errors = getValidationErrors(err);

        formRef.current?.setErrors(errors);
        return;
      }

      setLoading(false);

      Alert.alert(
        'Erro na autenticação',
        'Ocorreu um erro ao fazer login, cheque as credenciais'
      );
    }
  }, []);

  return (
    <>
      {loading && <Loading />}

      {!loading && (
        <>
          <Container>
            <Header>
              <Title>Login</Title>
            </Header>

            <Body>
              <Form ref={formRef} onSubmit={handleSignIn}>
                <Input
                  testID={testIDs.input_email}
                  autoCorrect={false}
                  autoCapitalize="none"
                  keyboardType="email-address"
                  name="email"
                  icon="mail"
                  placeholder="E-mail"
                  returnKeyType="next"
                  onSubmitEditing={() => {
                    passwordInputRef.current?.focus();
                  }}
                />

                <Input
                  testID={testIDs.input_password}
                  ref={passwordInputRef}
                  secureTextEntry
                  name="password"
                  icon="lock"
                  placeholder="Senha"
                  returnKeyType="send"
                  onSubmitEditing={() => {
                    formRef.current?.submitForm();
                  }}
                />

                <TouchableText isAlign={false} activeOpacity={0.6}>
                  <Text>Esqueceu sua senha ?</Text>
                  <Image source={forgotPassword} />
                </TouchableText>

                <Button
                  activeOpacity={0.6}
                  testID={testIDs.button}
                  onPress={() => {
                    formRef.current?.submitForm();
                  }}
                >
                  Entrar
                </Button>

                <TouchableText
                  onPress={() => navigation.navigate('/SignUp')}
                  isAlign
                  activeOpacity={0.6}
                >
                  <Text>
                    Não tem uma conta ? <Text isBold>Inscrever-se</Text>
                  </Text>
                </TouchableText>
              </Form>
            </Body>

            <Footer>
              <TextFooter>Ou faça login com sua conta</TextFooter>

              <Row>
                <TouchableOpacity activeOpacity={0.6}>
                  <Image source={iconGoogle} />
                </TouchableOpacity>

                <TouchableOpacity activeOpacity={0.6}>
                  <Image source={iconFacebook} />
                </TouchableOpacity>
              </Row>
            </Footer>
          </Container>
        </>
      )}
    </>
  );
};
