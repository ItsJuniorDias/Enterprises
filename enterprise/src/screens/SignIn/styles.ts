import styled from 'styled-components/native';
import { Colors, Fonts } from '../../theme';

type TouchableTextProps = {
  isAlign: boolean;
};

type TextProps = {
  isBold?: boolean;
};

export const Container = styled.View`
  flex: 1;
  padding: 16px 16px;
`;

export const Title = styled.Text`
  font-size: 28px;
  color: ${Colors.brand.secondary};
  font-family: ${Fonts.semiBold};
`;

export const Header = styled.View`
  margin-top: 32px;
`;

export const Body = styled.View`
  margin-top: 73px;
`;

export const TouchableText = styled.TouchableOpacity<TouchableTextProps>`
  width: 100%;
  flex-direction: row;
  justify-content: ${({ isAlign }) => (isAlign ? 'center' : 'flex-end')};
  margin-top: ${({ isAlign }) => (isAlign ? '24px' : '8px')};
`;

export const Text = styled.Text<TextProps>`
  font-size: 14px;
  color: #2d4379;
  font-family: 'Poppins-Regular';
  margin-right: 8px;
  font-weight: ${({ isBold }) => (isBold ? 700 : 400)};
`;

export const TextFooter = styled.Text`
  font-size: 14px;
  color: #2d4379;
  font-family: 'Poppins-Regular';
  margin-top: -8px;
  margin-bottom: 8px;
`;

export const Row = styled.View`
  flex-direction: row;
`;

export const Footer = styled.View`
  margin-top: 135px;
  width: 100%;
  align-items: center;
  justify-content: center;
`;
