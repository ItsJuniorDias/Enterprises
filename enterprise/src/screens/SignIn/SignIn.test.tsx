import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { SignIn } from '../SignIn/SignIn';
import { AxiosResponse } from 'axios';

const mockResponse = {
  data: {},
  status: 200,
} as AxiosResponse;

jest.mock('axios', () => ({
  create: jest.fn(() => ({
    post: jest.fn().mockImplementation(() => mockResponse),
  })),
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));
jest.mock('react-native-vector-icons/Feather', () => 'Icon');
jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

describe('Behavior SignIn', () => {
  const screenRender = () => <SignIn />;
  it('render snapshot', () => {
    const result = render(screenRender()).toJSON();

    expect(result).toMatchSnapshot();
  });

  it('should call function handleSignIn', async () => {
    const { getByTestId } = render(screenRender());

    const inputEmail = getByTestId('inputEmail_testID');

    fireEvent.changeText(inputEmail, 'juniordias@live.com');

    inputEmail.props.onSubmitEditing();

    const inputPassword = getByTestId('inputPassword_testID');

    fireEvent.changeText(inputPassword, '1234453');

    inputPassword.props.onSubmitEditing();

    const button = getByTestId('buttonSignIn_testID');

    fireEvent.press(button);

    //TODO expect
  });
});
