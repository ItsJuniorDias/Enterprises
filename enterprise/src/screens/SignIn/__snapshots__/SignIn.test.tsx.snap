// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Behavior SignIn render snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 0,
        "paddingLeft": 14,
        "paddingRight": 14,
        "paddingTop": 0,
      },
    ]
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "marginTop": 32,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#0d253c",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 28,
            },
          ]
        }
      >
        Login
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "marginTop": 73,
          },
        ]
      }
    >
      <View>
        <View
          isErrored={false}
          isFocused={false}
          isValue=""
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#f3f3f3",
                "borderColor": "#f3f3f3",
                "borderRadius": 10,
                "borderWidth": 2,
                "flexDirection": "row",
                "height": 60,
                "marginBottom": 8,
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "width": "100%",
              },
            ]
          }
          testID="container_inputId"
        >
          <Icon
            color="#9d9d9d"
            name="mail"
            size={20}
            style={
              Array [
                Object {
                  "marginRight": 16,
                },
              ]
            }
          />
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            autoCorrect={false}
            defaultValue=""
            keyboardAppearance="light"
            keyboardType="email-address"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSubmitEditing={[Function]}
            placeholder="E-mail"
            placeholderTextColor="#9d9d9d"
            rejectResponderTermination={true}
            returnKeyType="next"
            style={
              Array [
                Object {
                  "color": "#9d9d9d",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 16,
                },
              ]
            }
            testID="inputEmail_testID"
            underlineColorAndroid="transparent"
          />
        </View>
        <View
          isErrored={false}
          isFocused={false}
          isValue=""
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#f3f3f3",
                "borderColor": "#f3f3f3",
                "borderRadius": 10,
                "borderWidth": 2,
                "flexDirection": "row",
                "height": 60,
                "marginBottom": 8,
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "width": "100%",
              },
            ]
          }
          testID="container_inputId"
        >
          <Icon
            color="#9d9d9d"
            name="lock"
            size={20}
            style={
              Array [
                Object {
                  "marginRight": 16,
                },
              ]
            }
          />
          <TextInput
            allowFontScaling={true}
            defaultValue=""
            keyboardAppearance="light"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Senha"
            placeholderTextColor="#9d9d9d"
            rejectResponderTermination={true}
            returnKeyType="send"
            secureTextEntry={true}
            style={
              Array [
                Object {
                  "color": "#9d9d9d",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 16,
                },
              ]
            }
            testID="inputPassword_testID"
            underlineColorAndroid="transparent"
          />
        </View>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "marginTop": 8,
              "opacity": 1,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2d4379",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 14,
                  "marginRight": 8,
                },
              ]
            }
          >
            Esqueceu sua senha ?
          </Text>
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/round-arrow_right_alt-24px.png",
              }
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#376aed",
              "borderRadius": 10,
              "height": 60,
              "justifyContent": "center",
              "marginTop": 54,
              "opacity": 1,
              "width": "100%",
            }
          }
          testID="buttonSignIn_testID"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Poppins-Medium",
                  "fontSize": 18,
                },
              ]
            }
          >
            Entrar
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginTop": 135,
            "width": "100%",
          },
        ]
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2d4379",
                "fontFamily": "Poppins-Regular",
                "fontSize": 14,
                "marginBottom": 8,
                "marginTop": -8,
              },
            ]
          }
        >
          Ou fa√ßa login com sua conta
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/Google.png",
              }
            }
          />
        </View>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/Facebook.png",
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
